subsys/mgmt/mcumgr/smp/src/client.c

```diff
@@ -0,0 +1,279 @@
+
+	ARG_UNUSED(work);
+
+	if (sys_slist_is_empty(&smp_client_data.cmd_list)) {
+		/* No more packet for Transport */
+		return;
+	}
+
+	smpt = smp_transport_get();
+
+	SYS_SLIST_FOR_EACH_CONTAINER_SAFE(&smp_client_data.cmd_list, entry, tmp, node) {
+		if (smpt && k_uptime_get() - entry->timestamp < 1000) {
+			continue;
+		} else if (smpt && entry->retry_cnt < 15) {
+			/* Increment reference for re-transmission */
+			entry->nb = net_buf_ref(entry->nb);
+			entry->retry_cnt++;
+			entry->timestamp = k_uptime_get();
+			smp_tx_req(smpt, entry->nb);
+			continue;
+		}
+
+		cb = entry->cb;
+		user_data = entry->user_data;
+		smp_client_cmd_req_free(entry);
+		if (cb) {
+			cb(NULL, user_data);
+		}
+	}
+
+	if (!sys_slist_is_empty(&smp_client_data.cmd_list)) {
+		/* Re-schedule new timeout check*/
+		k_work_reschedule(&smp_client_data.work_delay, K_MSEC(50));
+	}
+}
+
+static int smp_client_init(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+	k_work_init_delayable(&smp_client_data.work_delay, snmp_client_trasnport_work_fn);
+	sys_slist_init(&smp_client_data.cmd_list);
+	sys_slist_init(&smp_client_data.cmd_free_list);
+	for (int i = 0; i < SMP_CMD_MAX; i++) {
+		sys_slist_append(&smp_client_data.cmd_free_list, &smp_cmd_req_bufs[i].node);
+	}
+	return 0;
+}
+
+static struct smp_client_cmd_req *smp_client_cmd_req_allocate(void)
+{
+	sys_snode_t *cmd_node;
+	struct smp_client_cmd_req *req;
+
+	cmd_node = sys_slist_get(&smp_client_data.cmd_free_list);
+	if (!cmd_node) {
+		return NULL;
+	}
+
+	req = SYS_SLIST_CONTAINER(cmd_node, req, node);
+
+	return req;
+}
+
+static void smp_cmd_add_to_list(struct smp_client_cmd_req *cmd_req)
+{
+	if (sys_slist_is_empty(&smp_client_data.cmd_list)) {
+		/* Enable timer */
+		k_work_reschedule(&smp_client_data.work_delay, K_MSEC(100));
+	}
+	sys_slist_append(&smp_client_data.cmd_list, &cmd_req->node);
+}
+
+static void smp_client_cmd_req_free(struct smp_client_cmd_req *cmd_req)
+{
+	smp_client_buf_free(cmd_req->nb);
+	cmd_req->nb = NULL;
+	sys_slist_find_and_remove(&smp_client_data.cmd_list, &cmd_req->node);
+	/* Add to free list*/
+	sys_slist_append(&smp_client_data.cmd_free_list, &cmd_req->node);
+
+	if (sys_slist_is_empty(&smp_client_data.cmd_list)) {
+		/* cancel delay*/
+		k_work_cancel_delayable(&smp_client_data.work_delay);
+	}
+}
+
+static struct smp_client_cmd_req *smp_client_response_discover(const struct smp_hdr *res_hdr)
+{
+	struct smp_hdr smp_header;
+	enum mcumgr_op_t response;
+	struct smp_client_cmd_req *cmd_req;
+
+	if (sys_slist_is_empty(&smp_client_data.cmd_list)) {
+		return NULL;
+	}
+
+	SYS_SLIST_FOR_EACH_CONTAINER(&smp_client_data.cmd_list, cmd_req, node) {
+		smp_read_hdr(cmd_req->nb, &smp_header);
+		if (smp_header.nh_op == MGMT_OP_READ) {
+			response = MGMT_OP_READ_RSP;
+		} else {
+			response = MGMT_OP_WRITE_RSP;
+		}
+
+		if (smp_header.nh_seq != res_hdr->nh_seq) {
+			continue;
+		} else if (res_hdr->nh_op != response) {
+			continue;
+		}
+
+		return cmd_req;
+	}
+	return NULL;
+}
```

Comment:
Not really sure what this is doing?

Reply:
It will discover proper request entry for received message